libz-rs-sys/src/gz.rs

```diff
@@ -607,6 +707,228 @@ pub unsafe extern "C-unwind" fn gzeof(file: gzFile) -> c_int {
+pub unsafe extern "C-unwind" fn gzdirect(file: gzFile) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return 0;
+    };
+
+    // In write mode, the direct flag was set in `gzopen_help`. In read mode, the
+    // direct status is determined lazily on the first read operation. If the first
+    // read hasn't happened yet, we can look at the header now to determine if the
+    // file is in gzip format.
+    if state.mode == GzMode::GZ_READ && state.how == How::Look && state.have == 0 {
+        _ = unsafe { gz_look(state) };
+    }
+
+    state.direct as _
+}
+
+// Given a gzip file opened for reading, check for a gzip header, and set
+// `state.direct` accordingly.
+//
+// # Returns
+//
+// * `Ok` on success
+// * `Err` on failure
+//
+// # Safety
+//
+// `state` must have been properly initialized, e.g. by [`gzopen_help`].
+unsafe fn gz_look(state: &mut GzState) -> Result<(), ()> {
+    // Allocate buffers if needed.
+    if state.input.is_null() {
+        state.in_size = state.want;
+        let Some(input) = ALLOCATOR.allocate_slice_raw::<u8>(state.in_size) else {
+            // Safety:
+            unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+            return Err(());
+        };
+        state.input = input.as_ptr();
+    }
+    if state.output.is_null() {
+        state.out_size = state.want * 2;
+        let Some(output) = ALLOCATOR.allocate_slice_raw::<u8>(state.out_size) else {
+            // Safety: The caller confirmed the validity of state, and free_buffers checks
+            // for null input and output pointers internally.
+            unsafe { free_buffers(state) };
+            // Safety: The caller confirmed the validity of state.
+            unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+            return Err(());
+        };
+        state.output = output.as_ptr();
+    }
+
+    // Allocate memory for inflate.
+    state.stream.avail_in = 0;
+    state.stream.next_in = ptr::null_mut();
+    // Safety: `gzopen_help` initialized `state.stream`'s `zalloc`, `zfree`, and
+    // `opaque` fields as needed by `inflateInit2`.
+    if unsafe { inflateInit2(&mut state.stream as *mut z_stream, MAX_WBITS + 16) } != Z_OK {
+        // Safety: The caller confirmed the validity of `state`, and `free_buffers` checks
+        // for null input and output pointers internally.
+        unsafe { free_buffers(state) };
+        // Safety: The caller confirmed the validity of `state`.
+        unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+        return Err(());
+    }
+
+    // Get at least the magic bytes in the input buffer.
+    if state.stream.avail_in < 2 {
+        // Safety: The caller confirmed the validity of `state`.
+        let Ok(_) = (unsafe { gz_avail(state) }) else {
+            return Err(());
+        };
+        if state.stream.avail_in == 0 {
```

Comment:
this could use the return value from `gz_avail`. Let's do that in this case actually, so you get `if gz_avail(state)? == 0` and so on.